src/main/kotlin/org/eln2/libelectric/sim/electrical/mna/impl/MnaCircuitImpl.kt

```diff
@@ -0,0 +1,103 @@
+package org.eln2.libelectric.sim.electrical.mna.impl
+
+import org.eln2.libelectric.sim.electrical.generic.*
+import org.eln2.libelectric.sim.electrical.mna.Circuit
+import org.eln2.libelectric.sim.electrical.mna.component.*
+
+class MnaCircuitImpl() : GenericCircuit {
+    var circuit = Circuit()
+    var components = mutableListOf<GenericComponent>()
+    var connectivity = mutableListOf<ComponentConnectivityMapping>()
+    override val backingImplementation = "ModifiedNodalAnalysis"
+
+    override fun addComponent(c: GenericComponent) {
+        components.add(c)
+        internalAddComponent(c)
+    }
+
+    private fun internalAddComponent(c: GenericComponent) {
+        return when (c) {
+            is GenericResistor -> {
+                val resistor = Resistor()
+                resistor.resistance = c.resistanceOhms
+                circuit.add(resistor)
+                c.internal = resistor
+            }
+            is GenericCapacitor -> {
+                val capacitor = Capacitor()
+                capacitor.capacitance = c.capacityFarads
+                capacitor.charge = 0.0
+                circuit.add(capacitor)
+                c.internal = capacitor
+            }
+            is GenericInductor -> {
+                val inductor = Inductor()
+                inductor.inductance = c.inductanceHenrie
+                inductor.flux = 0.0
+                circuit.add(inductor)
+                c.internal = inductor
+            }
+            is GenericVoltageSource -> {
+                val voltageSource = VoltageSource()
+                voltageSource.potential = c.voltage
+                circuit.add(voltageSource)
+                c.internal = voltageSource
+            }
+            is GenericCurrentSource -> {
+                val currentSource = CurrentSource()
+                currentSource.current = c.current
+                circuit.add(currentSource)
+                c.internal = currentSource
+            }
+            else -> {
+            }
```

Comment:
Is there a good way to abstract over this to save a few lines? This seems like a bit of boilerplate otherwise. (Maybe an in-scope function with a type parameter of the underlying component...)

Also, it really _is_ an error if that `else` is hit, because it means that something was well and truly not implemented and will have no effect.

Reply:
tbh, like 80% of this could be reduced by simply making the various constructors take optional arguments with the values intended. The `else->` is basically a catch statement for if it isn't any of those, eg a different type of component that is specific to a particular implementation.

Reply:
A default value would be opinionated, even if it's something otherwise simple (like `1`). Better idea, if at all possible:

```kotlin
private fun internalAddComponent(c: GenericComponent) {
  inline fun make<reified C: Component>(then: C.() -> Unit) {
    C().also { circuit.add(it) }.run(then)
  }
  when(c) {
    is GenericResistor -> make<Resistor> { resistance = c.resistanceOhms }
    is GenericCapacitor -> make<Capacitor> { capacitance = c.capacityFarads }
    /* etc. */
  }
}
```

I'm not confident enough in my Kotlin to know that'll work, but that looks much more straightforward to me _if_ it will :)